evennia/utils/evstring.py

```diff
@@ -0,0 +1,1497 @@
+        return ( EvLink(result[0], link_value=self._link_string, link_key=self._link_key), result[1], result[2] )
+
+    def rpartition(self, sep):
+        """
+        Splits the link text on the last instance of sep. Returns a tuple of length 3 containing a link
+        with the first segment of text, the separator, and an EvString of the tail. If sep is not found,
+        returns a tuple of two empty strings followed by self.
+        """
+        if sep not in self.clean():
+            return ('', '', self)
+        
+        result = self.ev_string.partition(sep, reverse=True)
+        return ( EvLink(result[0], link_value=self._link_string, link_key=self._link_key), result[1], result[2] )
+
+    def data(self):
+        """Returns the important elements of the link as a namedtuple"""
+        return LinkData(self._ev_string, self._link_string, self._link_key)
+    
+    def raw(self):
+        """Returns the original string with Evennia markup"""
+        return self._raw_string
+    
+    def clean(self):
+        """Returns just the text component of the link with no markup"""
+        return self._ev_string.clean()
+
+
+class EvString(str, metaclass=EvStringMeta):
+    """
+    Unicode-like object that is aware of Evennia format codes.
+
+    This class can be used nearly identically to strings, in that it will
+    report string length, handle slices, etc, much like a string object
+    would. The methods should be used identically as string methods are.
+
+    However, there are a few situations where the EvString will be converted
+    back to a raw markup string. When using ''.join() or u''.join() on an
+    EvString, or when using a regex method such as re.sub or re.search, Python
+    forces the EvString back to a simple string - you will have to convert the
+    results of those operations back to EvString.
+
+    """
+
+    # A compiled Regex for the format mini-language:
+    # https://docs.python.org/3/library/string.html#formatspec
+    re_format = re.compile(
+        r"(?i)(?P<just>(?P<fill>.)?(?P<align>\<|\>|\=|\^))?(?P<sign>\+|\-| )?(?P<alt>\#)?"
+        r"(?P<zero>0)?(?P<width>\d+)?(?P<grouping>\_|\,)?(?:\.(?P<precision>\d+))?"
+        r"(?P<type>b|c|d|e|E|f|F|g|G|n|o|s|x|X|%)?"
+    )
+
+    def __new__(cls, *args, **kwargs):
+        """
+        (this docstring is largely irrelevant for my target functionality: revisit)
+        When creating a new EvString, you may use a custom parser that has
+        the same attributes as the standard one, and you may declare the
+        string to be handled as already decoded. It is important not to double
+        decode strings, as escapes can only be respected once.
+
+        Internally, EvString can also passes itself precached code/character
+        indexes and clean strings to avoid doing extra work when combining
+        EvStrings.
+
+        """
+        text = args[0]
+        if not isinstance(text, str):
+            text = to_str(text)
```

Comment:
`to_str()` strips MXP from EvLinks, unlike EvCode, because it converts them to strings, so MXP will not survive being added to an EvString, e.g. by joining chunks. Perhaps EvLink should have a `__str__` function which returns `self.raw()` to avoid this?